File: backend/app/services/ics.py
"""Generates a standards-compliant .ics calendar from selected matches.

Design note: event times are written in UTC (from each match's kickoff_utc).
For fixed, one-time events this is the most portable choice — every calendar
client converts UTC to the viewer's own local timezone automatically, so the
kickoff shows correctly for every user with no VTIMEZONE blocks required. The
venue's local kickoff time is included in the description for reference.
"""
from __future__ import annotations

from datetime import datetime, timedelta, timezone

from icalendar import Calendar, Event

from ..models import Match, Team, Venue

PRODID = "-//worldcup-calendar-sync//FIFA World Cup 2026//EN"

# Emoji for flags that aren't a plain ISO 3166-1 alpha-2 country (UK home nations
# use Unicode subdivision tag sequences).
_SUBDIVISION_FLAGS = {
    "gb-eng": "\U0001F3F4\U000E0067\U000E0062\U000E0065\U000E006E\U000E0067\U000E007F",
    "gb-sct": "\U0001F3F4\U000E0067\U000E0062\U000E0073\U000E0063\U000E0074\U000E007F",
    "gb-wls": "\U0001F3F4\U000E0067\U000E0062\U000E0077\U000E006C\U000E0073\U000E007F",
}


def flag_emoji(team: Team) -> str:
    """Emoji flag for a team's country code, or '' for placeholders/unknowns."""
    code = team.code
    if not code:
        return ""
    if code in _SUBDIVISION_FLAGS:
        return _SUBDIVISION_FLAGS[code]
    if len(code) == 2 and code.isalpha():
        return "".join(chr(0x1F1E6 + ord(c) - ord("a")) for c in code.lower())
    return ""


def _team_label(team: Team) -> str:
    emoji = flag_emoji(team)
    return f"{emoji} {team.name}" if emoji else team.name


def event_summary(match: Match) -> str:
    return f"{match.stage.value}: {_team_label(match.home)} vs {_team_label(match.away)}"


def _stage_line(match: Match) -> str:
    if match.group:
        line = f"{match.stage.value} - Group {match.group}"
        if match.matchday:
            line += f" (Matchday {match.matchday})"
        return line
    return match.stage.value


def event_description(match: Match, venue: Venue) -> str:
    lines = [
        f"Stage: {_stage_line(match)}",
        f"Fixture: Match {match.match_number}",
        f"Venue: {venue.stadium}, {venue.city}, {venue.country}",
        f"Capacity: {venue.capacity:,}",
        f"Kickoff (venue local time): {match.kickoff_local.strftime('%a %d %b %Y, %H:%M')}",
    ]
    # Streaming/broadcast info will be appended here in a later milestone.
    return "\n".join(lines)


def event_location(venue: Venue) -> str:
    return f"{venue.stadium}, {venue.city}, {venue.country}"


def build_calendar(
    matches: list[Match],
    venues: dict[str, Venue],
    duration_minutes: int = 120,
    calname: str = "FIFA World Cup 2026",
) -> bytes:
    cal = Calendar()
    cal.add("prodid", PRODID)
    cal.add("version", "2.0")
    cal.add("method", "PUBLISH")
    cal.add("x-wr-calname", calname)

    stamp = datetime.now(timezone.utc)

    for match in matches:
        venue = venues[match.venue_id]

        start = match.kickoff_utc
        if start.tzinfo is None:
            start = start.replace(tzinfo=timezone.utc)
        end = start + timedelta(minutes=duration_minutes)

        event = Event()
        event.add("uid", f"wc2026-match-{match.match_number}@worldcup-calendar-sync")
        event.add("summary", event_summary(match))
        event.add("dtstart", start)
        event.add("dtend", end)
        event.add("dtstamp", stamp)
        event.add("location", event_location(venue))
        event.add("description", event_description(match, venue))
        cal.add_component(event)

    return cal.to_ical()

File: frontend/src/ics.ts
import type { Match, Team, Venue } from './types'

// Emoji for UK home nations (Unicode subdivision tag sequences).
const SUBDIVISION_FLAGS: Record<string, string> = {
  'gb-eng': '\u{1F3F4}\u{E0067}\u{E0062}\u{E0065}\u{E006E}\u{E0067}\u{E007F}',
  'gb-sct': '\u{1F3F4}\u{E0067}\u{E0062}\u{E0073}\u{E0063}\u{E0074}\u{E007F}',
  'gb-wls': '\u{1F3F4}\u{E0067}\u{E0062}\u{E0077}\u{E006C}\u{E0073}\u{E007F}',
}

function flagEmoji(team: Team): string {
  const code = team.code
  if (!code) return ''
  if (SUBDIVISION_FLAGS[code]) return SUBDIVISION_FLAGS[code]
  if (/^[a-z]{2}$/i.test(code)) {
    return [...code.toLowerCase()]
      .map((c) => String.fromCodePoint(0x1f1e6 + c.charCodeAt(0) - 97))
      .join('')
  }
  return ''
}

function teamLabel(team: Team): string {
  const emoji = flagEmoji(team)
  return emoji ? `${emoji} ${team.name}` : team.name
}

function eventSummary(m: Match): string {
  return `${m.stage}: ${teamLabel(m.home)} vs ${teamLabel(m.away)}`
}

function stageLine(m: Match): string {
  if (m.group) {
    let line = `${m.stage} - Group ${m.group}`
    if (m.matchday) line += ` (Matchday ${m.matchday})`
    return line
  }
  return m.stage
}

/** Venue wall-clock kickoff, e.g. "Thu 11 Jun 2026, 13:00". */
function localKickoffLabel(m: Match): string {
  // kickoff_local is naive venue time; read it as UTC so components don't shift.
  const d = new Date(`${m.kickoff_local}Z`)
  return new Intl.DateTimeFormat('en-GB', {
    weekday: 'short',
    day: '2-digit',
    month: 'short',
    year: 'numeric',
    hour: '2-digit',
    minute: '2-digit',
    timeZone: 'UTC',
  }).format(d)
}

function eventDescription(m: Match, v?: Venue): string {
  const lines = [`Stage: ${stageLine(m)}`, `Fixture: Match ${m.match_number}`]
  if (v) {
    lines.push(`Venue: ${v.stadium}, ${v.city}, ${v.country}`)
    lines.push(`Capacity: ${v.capacity.toLocaleString('en-US')}`)
  }
  lines.push(`Kickoff (venue local time): ${localKickoffLabel(m)}`)
  return lines.join('\n')
}

function eventLocation(v: Venue): string {
  return `${v.stadium}, ${v.city}, ${v.country}`
}

// Escape per RFC 5545 text rules.
function esc(s: string): string {
  return s.replace(/\\/g, '\\\\').replace(/;/g, '\\;').replace(/,/g, '\\,').replace(/\n/g, '\\n')
}

function icsStamp(d: Date): string {
  const p = (n: number) => String(n).padStart(2, '0')
  return (
    `${d.getUTCFullYear()}${p(d.getUTCMonth() + 1)}${p(d.getUTCDate())}` +
    `T${p(d.getUTCHours())}${p(d.getUTCMinutes())}${p(d.getUTCSeconds())}Z`
  )
}

/**
 * Build a VCALENDAR string. Event times are UTC so every client renders them in
 * the viewer's local zone. `reminderMinutes` (null = none, 0 = at kickoff) adds a
 * VALARM that fires that many minutes before kickoff.
 */
export function buildIcs(
  matches: Match[],
  venues: Map<string, Venue>,
  durationMin: number,
  reminderMinutes: number | null,
): string {
  const lines = [
    'BEGIN:VCALENDAR',
    'VERSION:2.0',
    'PRODID:-//worldcup-calendar-sync//FIFA World Cup 2026//EN',
    'METHOD:PUBLISH',
    'X-WR-CALNAME:FIFA World Cup 2026',
  ]
  const stamp = icsStamp(new Date())

  for (const m of matches) {
    const v = venues.get(m.venue_id)
    const start = new Date(m.kickoff_utc)
    const end = new Date(start.getTime() + durationMin * 60_000)

    lines.push('BEGIN:VEVENT')
    lines.push(`UID:wc2026-match-${m.match_number}@worldcup-calendar-sync`)
    lines.push(`SUMMARY:${esc(eventSummary(m))}`)
    lines.push(`DTSTART:${icsStamp(start)}`)
    lines.push(`DTEND:${icsStamp(end)}`)
    lines.push(`DTSTAMP:${stamp}`)
    if (v) lines.push(`LOCATION:${esc(eventLocation(v))}`)
    lines.push(`DESCRIPTION:${esc(eventDescription(m, v))}`)
    if (reminderMinutes !== null) {
      lines.push(
        'BEGIN:VALARM',
        'ACTION:DISPLAY',
        `DESCRIPTION:${esc(eventSummary(m))}`,
        `TRIGGER:-PT${reminderMinutes}M`,
        'END:VALARM',
      )
    }
    lines.push('END:VEVENT')
  }

  lines.push('END:VCALENDAR')
  return lines.join('\r\n') + '\r\n'
}

/** Trigger a browser download of an .ics string. */
export function downloadIcs(filename: string, content: string): void {
  const blob = new Blob([content], { type: 'text/calendar;charset=utf-8' })
  const url = URL.createObjectURL(blob)
  const a = document.createElement('a')
  a.href = url
  a.download = filename
  document.body.appendChild(a)
  a.click()
  a.remove()
  URL.revokeObjectURL(url)
}
